elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+         *
+         * Given that L and R operands of an OrFilterExpression can only be one of "pure-H", "pure-W", or "mix-HW",
+         * then:
+         *
+         *     pure-W1 OR pure-W2 = WHERE C1 OR WHERE C2 = WHERE (C1 OR C2)    = pure-W
+         *     pure-H1 OR pure-H2 = HAVING C1 OR HAVING C2 = HAVING (C1 OR C2) = pure-H
+         *
+         *     pure-H1 OR pureW2 = HAVING C1 OR WHERE C2 = HAVING (C1 OR C2)   = pure-H
+         *     pure-W1 OR pureH2 = WHERE C1 OR HAVING C2 = HAVING (C1 OR C2)   = pure-H
+         *
+         *     mix-HW1 OR pure-W2 = (WHERE C1 HAVING C1') OR WHERE C2 = HAVING (C1 & C1' | C2)  = pure-H
+         *     mix-HW1 OR pure-H2 = (WHERE C1 HAVING C1') OR HAVING C2 = HAVING (C1 & C1' | C2) = pure-H
+         *
+         *     mix-HW1 OR mim-HW2 = (WHERE C1 HAVING C1') OR (WHERE C2 HAVING C2') = HAVING ((C1 & C1') | (C2 & C2'))
+         *                        = pure-H
+         */
+
+        FilterConstraints left = expression.getLeft().accept(this);
+        FilterConstraints right = expression.getRight().accept(this);
+
+        if (left.isPureWhere() && right.isPureWhere()) {
+            // pure-W1 OR pure-W2 = WHERE (C1 OR C2) = pure-W
+            return FilterConstraints.pureWhere(
+                    OrFilterExpression.fromPair(
+                            left.getWhereExpression(),
+                            right.getWhereExpression()
+                    )
+            );
+        }  else {
+            // all of the rests are pure-H
+            return FilterConstraints.pureHaving(
+                    OrFilterExpression.fromPair(
+                            AndFilterExpression.fromPair(
+                                    left.getWhereExpression(),
+                                    left.getHavingExpression()
+                            ),
+                            AndFilterExpression.fromPair(
+                                    right.getWhereExpression(),
+                                    right.getHavingExpression()
+                            )
+                    )
+            );
+        }
+    }
+
+    @Override
+    public FilterConstraints visitNotExpression(NotFilterExpression expression) {
+        FilterExpression normalized = getNormalizationVisitor().visitNotExpression(expression);
+
+        if (normalized instanceof AndFilterExpression) {
+            return visitAndExpression((AndFilterExpression) normalized);
+        } else if (normalized instanceof OrFilterExpression) {
+            return visitOrExpression((OrFilterExpression) normalized);
+        } else if (normalized instanceof NotFilterExpression) {
+            FilterConstraints negatedConstraint = visitNotExpression((NotFilterExpression) normalized);
+
+            if (negatedConstraint.isPureWhere()) {
+                return FilterConstraints.pureWhere(new NotFilterExpression(negatedConstraint.getWhereExpression()));
+            } else if (negatedConstraint.isPureHaving()) {
```

Comment:
It isn't possible to have a mixed where/having for a NotFilterExpression after normalization.  You can just do a simple if/else here.